File: src/main/java/raylras/zen/code/type/IntRangeType.java
package raylras.zen.code.type;

import raylras.zen.code.SymbolProvider;
import raylras.zen.code.symbol.Operator;
import raylras.zen.code.symbol.Symbol;

import java.util.Collection;
import java.util.function.UnaryOperator;

public class IntRangeType extends Type implements SymbolProvider {

    public static final IntRangeType INSTANCE = new IntRangeType();

    @Override
    public String toString() {
        return "stanhebben.zenscript.value.IntRange";
    }

    @Override
    public Collection<Symbol> getSymbols() {
        return MembersBuilder.of()
                .variable("from", IntType.INSTANCE, Symbol.Modifier.VAL)
                .variable("to", IntType.INSTANCE, Symbol.Modifier.VAL)
                .operator(Operator.ITERATOR, new ListType(IntType.INSTANCE), UnaryOperator.identity())
                .build();
    }
}

File: core/src/main/kotlin/raylras/intellizen/type/IntRangeType.kt
package raylras.intellizen.type

import raylras.intellizen.CompilationEnvironment
import raylras.intellizen.symbol.Operator
import raylras.intellizen.symbol.Symbol
import raylras.intellizen.symbol.SymbolProvider
import raylras.intellizen.symbol.symbolSequence

object IntRangeType : Type, SymbolProvider {
    override val typeName = "stanhebben.zenscript.value.IntRange"

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        return symbolSequence {
            variable("from", IntType)
            variable("to", IntType)
            operator(Operator.FOR_IN, ListType(IntType))
        }
    }

    override fun toString() = typeName
}
